### ISSUE TITLE
Overriding pinned artifacts from another pin file

### ISSUE THREAD

**Issue by liucijus** [bazelbuild/rules_jvm_external#398]
I'm trying to find a way how to implement pinned artifact overrides. Currently, we have pinned deps in the central repo, and other repos load those pin files. But sometimes there's a need to override some of the centrally defined versions or add a repo-only dependency. Isolated artifact version trees are not sufficient for us, as we want to prevent the possibility of having the same artifact with different versions on a classpath. Having something like this would solve it:
```
maven_install(
   name = "merged_deps",
   maven_install_json = "@managed_maven_deps//:maven_install.json",
   override_maven_install_json = "@localrepo_maven_deps//:maven_install.json",
)
```
Specifying `override_maven_install_json` would mean to update `maven_install_json` content with deps and versions from the override file.

I'm currently thinking of implementing such merging within a custom repository_rule, but it would be great if such feature lands in `rules_jvm_external`.

**Comment by jin** [comment]
Alternative: In the local repo, why not merge the two *lists* of `artifacts` you're interested in, apply `version_conflict_policy` to `pinned`, then run artifact pinning to generate the "merged" json file. Does that work?

**Comment by liucijus** [comment]
I thought about doing it this way, but decided that working directly with pin files overcomes some challenges like:
 - how to update managed deps repo without having an additional step to regenerate pins in the rest of repos. It's not clear who and when should do it. 
 - reproducable control of dep closure generated by `artifacts = MANAGED_ARTIFACTS + OVERRIDEN_ARTIFACTS`.

Basically what I want to achieve, that pinning is done only if there's a change in top-level artifacts on the actual repo. Otherwise attempt an inexpensive merge of pin files.

**Comment by jin** [comment]
> how to update managed deps repo without having an additional step to regenerate pins in the rest of repos. It's not clear who and when should do it.

The managed deps repo can exposesthe list of artifacts that's loaded by child repos.

When the child repos update their version of the managed dep repo, run artifact pinning.

This should be reproducible.

**Comment by or-shachar** [comment]
Thanks @jin ! 

That idea would make sense if we update the version of managed dependencies repo occasionally and manually.

In our case - [we run in "virtual mono repository" mode](https://www.youtube.com/watch?v=2gNlTegwQD4) - meaning that for our own organization repositories - we update the commit hashes automatically multiple time. 

That's why we would like not to have additional steps performed after updating the managed dependencies version, especially not expensive ones. In general - generating the pin file from the artifacts is considered an expensive operation we wish to do as little as possible.

In our mind this is how we could solve it:
1. Managed deps repo generate pinned deps every time manage dep is updated.
2. Local repository generated pinned deps every time local dep is updated.
3. (a) **rules_jvm_external** `maven_install` would take both json files and merge them together into one json file.
   (b) alternatively - a custom repository rule would take both json files and merge them together, generating a single pin file to be consumed by the `maven_install` rule.

WDYT?

In general - do you think that merging multiple independently generated pinned file is feasible?

**Comment by jin** [comment]
> In general - do you think that merging multiple independently generated pinned file is feasible?

No, I don't wish to go down this path as the source of the JSON file should only be Coursier, and every maven_install should only read from maven_install.json. I'd like to keep the API simple. Merge semantics can make the maven_install rule complex really quickly.

I think a separate repository rule customized for your specific repository layout would be a better idea.
